rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp

```diff
@@ -708,3 +708,164 @@ TEST_F(TestNodeGraph, get_info_by_topic_endpoint_info_array_fini_error)
     node_graph()->get_publishers_info_by_topic("topic", false),
     rclcpp::exceptions::RCLError);
 }
+
+TEST_F(TestNodeGraph, get_action_names_and_types)
+{
+    ASSERT_LT(0u, node_graph()->get_action_names_and_types().size());
+}
+
+TEST_F(TestNodeGraph, get_action_names_and_types_rcl_error)
+{
+  auto mock = mocking_utils::patch_and_return(
+    "lib:rclcpp", rcl_action_get_names_and_types, RCL_RET_ERROR);
+    rclcpp::exceptions::RCLError);
+  auto mock_info_array_fini = mocking_utils::patch_and_return(
+    "lib:rclcpp", rcl_names_and_types_fini, RCL_RET_ERROR);
+  EXPECT_THROW(
+    node_graph()->get_action_names_and_types("action", false),
+    rclcpp::exceptions::RCLError);
+
+TEST_F(TestNodeGraph, get_action_names_and_types_rcl_fini_error)
+{
+  auto mock_info_array_fini = mocking_utils::patch_and_return(
+    "lib:rclcpp", rcl_names_and_types_fini, RCL_RET_ERROR);
+  EXPECT_THROW(
+    node_graph()->get_action_names_and_types("action", false),
+    rclcpp::exceptions::RCLError);
+}
+
+TEST_F(TestNodeGraph, get_action_client_names_and_types_by_node)
+{
+  auto client =
+    node()->create_client<test_msgs::srv::Empty>("node1_service");
+
+  const std::string node2_name = "node2";
+  auto node2 = std::make_shared<rclcpp::Node>(node2_name, node_namespace);
+
+  // rcl_action_get_client_names_and_types_by_node() expects the node to exist, otherwise it fails
+  EXPECT_THROW(
+    node_graph()->get_action_client_names_and_types_by_node("not_a_node", "not_absolute_namespace"),
+    std::runtime_error);
+
+  auto services_of_node1 =
+    node_graph()->get_action_client_names_and_types_by_node(node_name, absolute_namespace);
+  auto services_of_node2 =
+    node_graph()->get_action_client_names_and_types_by_node(node2_name, absolute_namespace);
+
+  // Begin microbenchmarking of the action names
+  auto start = std::chrono::steady_clock::now();
+  while (std::chrono::steady_clock::now() - start < std::chrono::seconds(3)) {
+    services_of_node1 =
+      node_graph()->get_action_client_names_and_types_by_node(node_name, absolute_namespace);
+    services_of_node2 =
+      node_graph()->get_action_client_names_and_types_by_node(node2_name, absolute_namespace);
+    if (services_of_node1.find("/ns/node1_service") != services_of_node1.end()) {
+      break;
+    }
+  }
+
+  EXPECT_TRUE(services_of_node1.find("/ns/node1_service") != services_of_node1.end());
+  EXPECT_FALSE(services_of_node2.find("/ns/node1_service") != services_of_node2.end());
+}
+
+TEST_F(TestNodeGraph, get_action_client_names_and_types_by_node_rcl_error)
+{
+  auto client = node()->create_client<test_msgs::srv::Empty>("node1_service");
```

Comment:
same here, this is creating `service` but `action`. `rclcpp_action::create_client` should be called instead.

Reply:
While this makes sense, I cannot include the rclcpp_action part of the library in the rclcpp section, otherwise a linking error is thrown because rclcpp_action uses rclcpp, so this would result in an infinite loop. We need to both, have these functions in the node_graph_interface section, and utilize the rclcpp_action commands.

Reply:
One thing that we could do is add specifically the tests to the `rclcpp_action` part of the `rclcpp` [library](https://github.com/ros2/rclcpp/tree/rolling/rclcpp_action/test) with a new file along the names of test_action_calls. Since rclcpp_action has to include rclcpp in its [cmake file](https://github.com/ros2/rclcpp/blob/rolling/rclcpp_action/CMakeLists.txt#L7) and already has the ability to include `rclcpp_action::create_server` there would be no linking issues. The problem is just the location of these files because the `rcl_action` wrapper interfaces are in the `rclcpp/nodeinterfaces` section of the codebase, which makes this location a bit awkward for naming purpose. The functions like `get_action_types_and_names` make the most sense where they are now, but it seems the tests have to be moved simply because including rclcpp_action in rclcpp isn't possible.

Reply:
This conversation is making me question whether we want to expose this as part of `rclcpp` at all.  That is, I wonder if it would make more sense to move all of this to `rclcpp_action` (possibly with a new `NodeActionGraph` class), which avoids this whole problem.  That said, I haven't looked at this implementation closely yet, so maybe this doesn't make much sense.

Reply:
I would have to agree with moving this entire implementation to rclcpp_action. The folder is already quite slim with files and seems to be a better place to place wrapped rcl_action functions. While keeping all three of these functions in the node_graph.cpp section would be simpler; the fact that we would need to move tests to a different part of the project entirely showcases a design flaw.

Reply:
I'm now less confident in moving the entire implementation, the NodeGraph monitors a node given as `rclcpp::node_interfaces::NodeBaseInterface` . If we're monitoring the one node, and that node can provide a range of topics, services, and actions, then we split this implementation into two pieces, would the user lose the ability to keep all of that data in one node? Or would we just have to pass that same `clcpp::node_interfaces::NodeBaseInterface` into the constructor of this proposed `NodeActionGraph`, in order to view all of this data (assuming the user is fine with creating both a `NodeGraph` and a `NodeActionGraph`

Reply:
I'm now less confident in moving the entire implementation of `NodeGraph` into `rclcpp_action` as `NodeActionGraph`. While we are only using it for action based activities, it's not action specific, simply just an rcl_action wrapper, we then have to move all of the other general overrides to this new class. This will add another ~500 lines that we probably could have avoided.

 Additionally, if the user has some long-running process using an action server/client, while they have a small publisher/subscriber maybe keeping track of the environment, they won't be able to track these processes from one `NodeGraph`. If we did move the action side of things over to a new class, they have to create both a `NodeGraph` and `NodeActionGraph`, so if the user were to call `get_graph_event` and [gain another user](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/node_interfaces/node_graph.cpp#L558) on the `NodeGraph`, the `NodeActionGraph` wouldn't see that change, thus won't be able to accurately track the processes running.